### PR TITLE
Allow getting logs directly from deployment, job and statefulset

### PR DESCRIPTION
**Pull request by soltysh**
**Special notes for your reviewer**:
@smarterclayton you asked for it in OpenShift

```release-note
kubectl logs allows getting logs directly from deployment, job and statefulset
```

### CODE REVIEW

**Comment by soltysh** [comment]
@kubernetes/sig-apps-pr-reviews

**Comment by pwittrock** [comment]
Do tests exist for these?

**Comment by soltysh** [comment]
@pwittrock I haven't seen any, I'll add one.

**Comment by pwittrock** [comment]
That would be wonderful.  Thank you.

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 1]
```diff
@@ -260,6 +260,26 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
...
+		opts, ok := options.(*api.PodLogOptions)
+		if !ok {
+			return nil, errors.New("provided options object is not a PodLogOptions")
+		}
+		selector, err := metav1.LabelSelectorAsSelector(t.Spec.Selector)
+		if err != nil {
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		sortBy := func(pods []*v1.Pod) sort.Interface { return controller.ByLogging(pods) }
+		pod, numPods, err := GetFirstPod(clientset.Core(), t.Namespace, selector, 20*time.Second, sortBy)
```
Comment (krmayankk) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Will this only work for the first pod ?

Comment (smarterclayton) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Yes, consistent with the others.  See https://github.com/kubernetes/kubernetes/issues/19873 for proposed extension.

**Comment by kargakis** [comment]
@soltysh I think the help message of the command needs an update

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 2]
```diff
@@ -260,6 +260,26 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
...
+		if err != nil {
+			return nil, err
+		}
+		if numPods > 1 {
+			fmt.Fprintf(os.Stderr, "Found %v pods, using pod/%v\n", numPods, pod.Name)
```
Comment (mfojtik) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: why is this error?

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 1 above)
Comment (mfojtik) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: shouldn't we check if replicas > 0 ? AFAIK this function will wait for 20s for a Pod to appear, but I think we should return faster

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 2 above)
Comment (soltysh) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Consistently with others.

**Comment by soltysh** [comment]
I've added more objects per @smarterclayton request, that is StatefulSet, Job and CronJob. I've also added them to `AttachablePodForObject`. I still need to figure out the test for this.

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 3]
```diff
@@ -330,6 +390,16 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object) (*api.Pod,
 		sortBy := func(pods []*v1.Pod) sort.Interface { return sort.Reverse(controller.ActivePods(pods)) }
 		pod, _, err := GetFirstPod(clientset.Core(), t.Namespace, selector, 1*time.Minute, sortBy)
 		return pod, err
+
+	case *extensions.ReplicaSet:
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: There is an open PR for attach, mind leaving that aside?
https://github.com/kubernetes/kubernetes/pull/40365

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: If it's consistent with the others, why would we change the behavior here?

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 3 above)
Comment (soltysh) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Sure thing.

**Comment by soltysh** [comment]
@kargakis I've removed the attach part
@pwittrock added tests, although now that we have a placeholder for more tests probably `AttachablePodForObject` could deserve its own tests as well :) (as a followup)
@deads2k and @fabianofranz  you might be interested in the first commit, for the `ClientAccessFactory` to be usable for tests I had to change method signatures to use `internalclientset.Interface` instead of actual type, can one of you approve that change?
@smarterclayton ready for final review

File: pkg/kubectl/cmd/logs.go [hunk 4]
```diff
@@ -54,13 +54,19 @@ var (
...
 
 		# Show all logs from pod nginx written in the last hour
-		kubectl logs --since=1h nginx`)
+		kubectl logs --since=1h nginx
+
+		# Return snapshot logs from first container of a job named hello
+		kubectl logs job/hello
+
+		# Return snapshot logs from from container hello-1 of a job named hello
+		kubectl logs job/hello -c hello-1`)
```
Comment (kargakis) [inline pkg/kubectl/cmd/logs.go]: Provide an example with a different resource since you already have one with jobs above

**Comment by kargakis** [comment]
The relase note needs an update

**Comment by kargakis** [comment]
Apart from a couple of comments and the clientset interface update, lgtm

**Comment by soltysh** [comment]
@kargakis nits fixed, release note updated

**Comment by soltysh** [comment]
@k8s-bot kubemark e2e test this

**Comment by shiywang** [comment]
hi @soltysh @kargakis just have a simple question here, since `logs` can support multiple type, does it support those formats below ?
``` console
kubectl logs rc rc-test 
kubectl logs rc rc-test containerName
```

File: pkg/kubectl/cmd/logs.go [hunk 5]
```diff
@@ -94,9 +100,7 @@ func NewCmdLogs(f cmdutil.Factory, out io.Writer) *cobra.Command {
 		},
 		Run: func(cmd *cobra.Command, args []string) {
 			cmdutil.CheckErr(o.Complete(f, out, cmd, args))
-			if err := o.Validate(); err != nil {
-				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
-			}
+			cmdutil.CheckErr(o.Validate())
```
Comment (shiywang) [inline pkg/kubectl/cmd/logs.go]: why remove this ? if a user provided an invalid logs options, shouldn't we prompt this ?
```console
unexpected logs options object 
See 'logs -h' for help and examples.
```

Comment (soltysh) [inline pkg/kubectl/cmd/logs.go]: All user input is verified in `Complete`, at this point we're validating the correctness of internal structures (here `LogsOptions`) before running the actual command. User has no real input on these values, thus this doesn't provide any value for him. See other commands for similarity.

Comment (shiywang) [inline pkg/kubectl/cmd/logs.go]: ok, I see, thanks

**Comment by soltysh** [comment]
> hi @soltysh @kargakis just have a simple question here, since logs can support multiple type, does it support those formats below ?
kubectl logs rc rc-test 
kubectl logs rc rc-test containerName

No, for these resources (jobs, statefulset, etc) you need to pass resource/name explicitly, to make it work.

**Comment by soltysh** [comment]
Rebased.

**Comment by kargakis** [comment]
Initially we couldn't support RESOURCE NAME syntax because of backwards
compatibility (second arg used to be the container name for the pod)

**Comment by soltysh** [comment]
@k8s-bot cvm gke e2e test this

**Comment by soltysh** [comment]
@k8s-bot kubemark e2e test this

**Comment by soltysh** [comment]
@pwittrock ptal

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 6]
```diff
@@ -271,6 +268,16 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
 		}
 		return nil, fmt.Errorf("cannot get the logs from %v", gvks[0])
 	}
+
+	sortBy := func(pods []*v1.Pod) sort.Interface { return controller.ByLogging(pods) }
+	pod, numPods, err := GetFirstPod(clientset.Core(), namespace, selector, 20*time.Second, sortBy)
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: We really need to make the timeout configurable. Spawn an issue?

Comment (soltysh) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: The same applies for `AttachablePodForObject`, I've created https://github.com/kubernetes/kubernetes/issues/41786 to track it.

**Comment by deads2k** [comment]
It's a minor refactor and the right choice for federation too.

**Comment by soltysh** [comment]
This is approved and lgtm, I'll apply the label manually.

**Comment by lukaszo** [comment]
Is there any reason to not include daemon sets here?

**Comment by fejta** [comment]
@k8s-bot kubemark e2e test this https://github.com/kubernetes/test-infra/issues/2012

**Comment by soltysh** [comment]
> Is there any reason to not include daemon sets here?

@lukaszo I remember having some doubts about Daemon Sets and Cron Jobs so I didn't added right now. And as long as CJ I'm quite positive it's hard to justify how to do it properly you're free to reason about Daemon Sets.

**Comment by soltysh** [comment]
@k8s-bot kops aws e2e test this

**Comment by soltysh** [comment]
Rebased.

**Comment by soltysh** [comment]
Re-applying lgtm label, since this was just mechanical rebase.

**Comment by soltysh** [comment]
@k8s-bot unit test this

**Comment by soltysh** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by fgrzadkowski** [comment]
@soltysh What's the motivation for this PR? How users should use it? AFAIU the semantic is that we will show logs from a random pod within Deployment. How is it useful? To be hones I find it pretty confusing.

**Comment by soltysh** [comment]
@fgrzadkowski this is more of a convenience when you have a simple deployment/job/stateful set. We've had similar logic in place for RC/RS already, this was only adding the same for other objects.

**Comment by fgrzadkowski** [comment]
I see. Though I find it very confusing. So we you saying that this is intended for Deployments/RS/RC with only one instance where it's clear what this mean? Do you know what was the motivation for label selectors?

**Comment by soltysh** [comment]
Not sure I understand which label selectors you're talking about, here. The ones used in this impl are only to get the pods owned by the controller and then pick the first pod.

**Comment by kargakis** [comment]
[The way](https://github.com/kubernetes/kubernetes/blob/9a0f5ccb331b6bb8dc3d6e2c91d05bd2af15c1c0/pkg/kubectl/cmd/util/factory_object_mapping.go#L291) pods are picked up [is not random](https://github.com/kubernetes/kubernetes/blob/9a0f5ccb331b6bb8dc3d6e2c91d05bd2af15c1c0/pkg/controller/controller_utils.go#L688) but we may want to document it.

**Comment by fgrzadkowski** [comment]
To be honest I find that logic even more confusing.

I was referring to the following option in `kubectl logs`:
```
  -l, --selector='': Selector (label query) to filter on.
```

Maybe it'd useful to at least let the user choose which pod they are interested in?

**Comment by soltysh** [comment]
See https://github.com/kubernetes/kubernetes/pull/44282 for the relevant discussion.

**Pull request merged**